ludwig/backend/horovod.py

```diff
@@ -83,3 +84,23 @@ def get_available_resources(self) -> Resources:
         gpus = self._horovod.allreduce(gpus, op=self._horovod.Sum).item()
 
         return Resources(cpus=cpus, gpus=gpus)
+
+    def get_max_concurrent_trials(self, hyperopt_config: Dict[str, Any]) -> Union[int, None]:
+        cpus_per_trial = hyperopt_config[EXECUTOR].get(CPU_RESOURCES_PER_TRIAL, 1)
+        num_cpus_available = self.get_available_resources().cpus
+
+        # no actors will compete for ray datasets tasks dataset tasks are cpu bound
+        if cpus_per_trial == 0:
+            return None
+
+        if num_cpus_available < 2:
+            raise RuntimeError("At least 2 CPUs are required for hyperopt when using a HorovodBackend.")
+
+        # Ray requires at least 2 free CPUs to ensure trials don't stall
+        max_possible_trials = (num_cpus_available - 2) // cpus_per_trial
+
+        # Users may be using an autoscaling cluster, so return None
+        if max_possible_trials < 1:
+            return None
+
+        return max_possible_trials
```

Comment:
Refactor to pull this function out for RayBackend and HorovodBackend to prevent code duplication